src/diff.js

```diff
@@ -0,0 +1,289 @@
+			let i;
+			// if rhs/lhs lines have been retained, output contextual diff
+			for (i = item.lhs_start; i < item.lhs_start + item.lhs_deleted_count; ++i) {
+				nf += '< ' + lhs_lines[i] + '\n';
+			}
+			if (item.rhs_inserted_count && item.lhs_deleted_count) nf += '---\n';
+			for (i = item.rhs_start; i < item.rhs_start + item.rhs_inserted_count; ++i) {
+				nf += '> ' + rhs_lines[i] + '\n';
+			}
+		}
+	}
+	return nf;
+};
+
+diff.prototype._lcs = function(lhs_ctx, lhs_lower, lhs_upper, rhs_ctx, rhs_lower, rhs_upper, vector_u, vector_d) {
+	while ( (lhs_lower < lhs_upper) && (rhs_lower < rhs_upper) && (lhs_ctx.codes[lhs_lower] == rhs_ctx.codes[rhs_lower]) ) {
+		++lhs_lower;
+		++rhs_lower;
+	}
+	while ( (lhs_lower < lhs_upper) && (rhs_lower < rhs_upper) && (lhs_ctx.codes[lhs_upper - 1] == rhs_ctx.codes[rhs_upper - 1]) ) {
+		--lhs_upper;
+		--rhs_upper;
+	}
+	if (lhs_lower == lhs_upper) {
+		while (rhs_lower < rhs_upper) {
+			rhs_ctx.modified[ rhs_lower++ ] = true;
+		}
+	}
+	else if (rhs_lower == rhs_upper) {
+		while (lhs_lower < lhs_upper) {
+			lhs_ctx.modified[ lhs_lower++ ] = true;
+		}
+	}
+	else {
+		const sms = this._sms(lhs_ctx, lhs_lower, lhs_upper, rhs_ctx, rhs_lower, rhs_upper, vector_u, vector_d);
+		this._lcs(lhs_ctx, lhs_lower, sms.x, rhs_ctx, rhs_lower, sms.y, vector_u, vector_d);
+		this._lcs(lhs_ctx, sms.x, lhs_upper, rhs_ctx, sms.y, rhs_upper, vector_u, vector_d);
+	}
+};
+
+diff.prototype._sms = function(lhs_ctx, lhs_lower, lhs_upper, rhs_ctx, rhs_lower, rhs_upper, vector_u, vector_d) {
+	const max = lhs_ctx.codes.length + rhs_ctx.codes.length + 1;
+	const kdown = lhs_lower - rhs_lower;
+	const kup = lhs_upper - rhs_upper;
+	const delta = (lhs_upper - lhs_lower) - (rhs_upper - rhs_lower);
+	const odd = (delta & 1) != 0;
+	const offset_down = max - kdown;
+	const offset_up = max - kup;
+	const maxd = ((lhs_upper - lhs_lower + rhs_upper - rhs_lower) / 2) + 1;
+	vector_d[ offset_down + kdown + 1 ] = lhs_lower;
+	vector_u[ offset_up + kup - 1 ] = lhs_upper;
+	const ret = { x:0, y:0 }
+	let x;
+	let y;
+	for (let d = 0; d <= maxd; ++d) {
+		for (let k = kdown - d; k <= kdown + d; k += 2) {
+			if (k == kdown - d) {
+				x = vector_d[ offset_down + k + 1 ];//down
+			}
+			else {
+				x = vector_d[ offset_down + k - 1 ] + 1;//right
+				if ((k < (kdown + d)) && (vector_d[ offset_down + k + 1 ] >= x)) {
+					x = vector_d[ offset_down + k + 1 ];//down
+				}
+			}
+			y = x - k;
+			// find the end of the furthest reaching forward D-path in diagonal k.
+			while ((x < lhs_upper) && (y < rhs_upper) && (lhs_ctx.codes[x] == rhs_ctx.codes[y])) {
+				x++; y++;
+			}
+			vector_d[ offset_down + k ] = x;
+			// overlap ?
+			if (odd && (kup - d < k) && (k < kup + d)) {
+				if (vector_u[offset_up + k] <= vector_d[offset_down + k]) {
+					ret.x = vector_d[offset_down + k];
+					ret.y = vector_d[offset_down + k] - k;
+					return (ret);
+				}
+			}
+		}
+		// Extend the reverse path.
+		for (k = kup - d; k <= kup + d; k += 2) {
```

Comment:
assignment to undeclared variable k
`for (let k = kup - d; k <= kup + d; k += 2) {`